### ISSUE TITLE
[YoloV8] Torch compile model shows metrics degradation on the coco128 dataset

### ISSUE THREAD

**Issue by daniil-lyakhov** [ultralytics/ultralytics#13800]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Val

### Bug
Yolov8 n model shows significant metrics degradation after the torch.compile:
```
Torch model:
FP32 torch model validation results:
val: Scanning /home/dlyakhov/Projects/nncf/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|████████
100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 128/128 [00:05<00:00, 21.41it/s]
               Class      Images      Labels   Precision      Recall      mAP@.5  mAP@.5:.95
                 all         128         929       0.651       0.532       0.606       0.452
Torch FX model:
FP32 FX model validation results:
val: Scanning /home/dlyakhov/Projects/nncf/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|████████
100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 128/128 [00:32<00:00,  3.90it/s]
               Class      Images      Labels   Precision      Recall      mAP@.5  mAP@.5:.95
                 all         128         929       0.524       0.145       0.151      0.0991
```

### Environment

Ultralytics YOLOv8.2.35 🚀 Python-3.8.10 torch-2.3.1+cu121 CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
Setup complete ✅ (36 CPUs, 125.5 GB RAM, 1176.2/1832.2 GB disk)

OS                  Linux-5.15.0-105-generic-x86_64-with-glibc2.29
Environment         Linux
Python              3.8.10
Install             pip
RAM                 125.48 GB
CPU                 Intel Core(TM) i9-10980XE 3.00GHz
CUDA                12.1

numpy               ✅ 1.24.3<2.0.0
matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.3.1>=1.8.0
torchvision         ✅ 0.18.1>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.0>=2.0.0

### Minimal Reproducible Example

```python
# torch==2.3.1
# ultralytics==8.2.35
from typing import Dict, Tuple

import numpy as np
import torch
from tqdm import tqdm
from ultralytics.data.utils import check_det_dataset
from ultralytics.engine.validator import BaseValidator as Validator
from ultralytics.models.yolo import YOLO
from ultralytics.utils.torch_utils import de_parallel


def print_statistics(stats: np.ndarray, total_images: int, total_objects: int) -> None:
    mp, mr, map50, mean_ap = (
        stats["metrics/precision(B)"],
        stats["metrics/recall(B)"],
        stats["metrics/mAP50(B)"],
        stats["metrics/mAP50-95(B)"],
    )
    s = ("%20s" + "%12s" * 6) % ("Class", "Images", "Labels", "Precision", "Recall", "mAP@.5", "mAP@.5:.95")
    print(s)
    pf = "%20s" + "%12i" * 2 + "%12.3g" * 4  # print format
    print(pf % ("all", total_images, total_objects, mp, mr, map50, mean_ap))


def prepare_validation(model: YOLO, data: str) -> Tuple[Validator, torch.utils.data.DataLoader]:
    custom = {"rect": False, "batch": 1}  # method defaults
    args = {**model.overrides, **custom, "mode": "val"}  # highest priority args on the right

    validator = model._smart_load("validator")(args=args, _callbacks=model.callbacks)
    stride = 32  # default stride
    validator.stride = stride  # used in get_dataloader() for padding
    validator.data = check_det_dataset(data)
    validator.init_metrics(de_parallel(model))

    data_loader = validator.get_dataloader(validator.data.get(validator.args.split), validator.args.batch)
    return validator, data_loader


def validate(model, data_loader: torch.utils.data.DataLoader, validator: Validator) -> Tuple[Dict, int, int]:
    with torch.no_grad():
        for batch in data_loader:
            batch = validator.preprocess(batch)
            preds = model(batch["img"])
            preds = validator.postprocess(preds)
            validator.update_metrics(preds, batch)
        stats = validator.get_stats()
    return stats, validator.seen, validator.nt_per_class.sum()


def main(torch_fx):
   yolo_model = YOLO("yolov8n")

    model_type = "torch"
    model = yolo_model.model
    if torch_fx:
        model = torch.compile(model)
        model_type = "FX"
    print(f"FP32 {model_type} model validation results:")
    validator, data_loader = prepare_validation(yolo_model, "coco128.yaml")
    stats, total_images, total_objects = validate(model, tqdm(data_loader), validator)
    print_statistics(stats, total_images, total_objects)


if __name__ == "__main__":
    print("Torch model:")
    main(torch_fx=False)
    print("Torch FX model:")
    main(torch_fx=True)
```

### Additional

It is not possible to run `yolo_model.val()` to check compiled model metrics as the `val` does not use optimized model inside the validation loop: `AutoBackend` object is using original module instead of optimized

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by daniil-lyakhov** [comment]
Minimal reproducer:
```python
# torch==2.3.1
# ultralytics==8.2.35
import torch
from ultralytics.models.yolo import YOLO


torch.manual_seed(42)

def run_yolo(torch_fx, inputs):
    yolo_model = YOLO("yolov8n")
    model = yolo_model.model
    if torch_fx:
        model = torch.compile(model)
    return model(inputs)[0]


if __name__ == "__main__":
    inputs = torch.rand((1, 3, 640, 640))
    print("Run Torch model...")
    torch_t = run_yolo(torch_fx=False, inputs=inputs)
    print("Run Torch FX model...")
    fx_t = run_yolo(torch_fx=True, inputs=inputs)

    abs_diff = torch.abs(torch_t - fx_t)
    idx = torch.argmax(abs_diff)
    print(f"argmax idx: {idx}")
    print(f"torch value: {torch_t.view(-1)[idx]}")
    print(f"torch FX value: {fx_t.view(-1)[idx]}")
    print(f'abs diff: {abs_diff.view(-1)[idx]}')
    print(f"torch.quantile(abs_diff, 0.96) {torch.quantile(abs_diff, 0.96)}")
```
```
Run Torch model...
Run Torch FX model...
argmax idx: 25132
torch value: 490.80194091796875
torch FX value: 855.9827270507812
abs diff: 365.1807861328125
torch.quantile(abs_diff, 0.96) 2.0144500732421875
```

**Comment by glenn-jocher** [comment]
@daniil-lyakhov hi there,

Thank you for providing the minimal reproducible example and detailed information about the issue you're encountering with the `torch.compile` model showing metrics degradation on the COCO128 dataset.

It appears that you've identified a significant difference in the validation metrics between the standard PyTorch model and the Torch FX compiled model. This discrepancy is indeed concerning and warrants further investigation.

### Steps to Investigate:

1. **Verify Versions**:
   Ensure you are using the latest versions of both `torch` and `ultralytics`. The versions you mentioned (`torch==2.3.1` and `ultralytics==8.2.35`) are quite recent, but it's always good to double-check for any new updates or patches that might address this issue.

2. **Model Consistency Check**:
   The minimal example you provided shows a significant difference in the output values between the standard and compiled models. This suggests that the compilation process might be altering the model's behavior. To further diagnose this, you can compare intermediate outputs (e.g., feature maps) at various layers of the model for both the standard and compiled versions. This can help pinpoint where the discrepancy begins.

3. **Validation Loop**:
   As you noted, the `val` method does not currently use the optimized model inside the validation loop. You can modify the validation loop to use the compiled model directly, ensuring that the same model is being evaluated:
   ```python
   def validate(model, data_loader: torch.utils.data.DataLoader, validator: Validator) -> Tuple[Dict, int, int]:
       with torch.no_grad():
           for batch in data_loader:
               batch = validator.preprocess(batch)
               preds = model(batch["img"])
               preds = validator.postprocess(preds)
               validator.update_metrics(preds, batch)
           stats = validator.get_stats()
       return stats, validator.seen, validator.nt_per_class.sum()
   ```

4. **Precision and Stability**:
   The differences in precision and stability between the standard and compiled models could be due to various factors, including numerical stability issues introduced during the compilation process. You might want to experiment with different compilation settings or flags provided by `torch.compile` to see if they mitigate the issue.

### Example Code for Validation with Compiled Model:
Here's an example of how you can modify the validation loop to use the compiled model:
```python
def main(torch_fx):
    yolo_model = YOLO("yolov8n")
    model_type = "torch"
    model = yolo_model.model
    if torch_fx:
        model = torch.compile(model)
        model_type = "FX"
    print(f"FP32 {model_type} model validation results:")
    validator, data_loader = prepare_validation(yolo_model, "coco128.yaml")
    stats, total_images, total_objects = validate(model, tqdm(data_loader), validator)
    print_statistics(stats, total_images, total_objects)
```

### Next Steps:
1. **Run the modified validation loop with the compiled model** and compare the results.
2. **Check for any updates** to `torch` and `ultralytics` that might address this issue.
3. **Experiment with different compilation settings** to see if they affect the model's performance and accuracy.

If the issue persists, please let us know, and we can further investigate potential causes and solutions.

Thank you for your patience and for bringing this to our attention. We look forward to resolving this issue with your help.

**Comment by daniil-lyakhov** [comment]
> @daniil-lyakhov hi there,
> 
> Thank you for providing the minimal reproducible example and detailed information about the issue you're encountering with the `torch.compile` model showing metrics degradation on the COCO128 dataset.
> 
> It appears that you've identified a significant difference in the validation metrics between the standard PyTorch model and the Torch FX compiled model. This discrepancy is indeed concerning and warrants further investigation.
> 
> ### Steps to Investigate:
> 1. **Verify Versions**:
>    Ensure you are using the latest versions of both `torch` and `ultralytics`. The versions you mentioned (`torch==2.3.1` and `ultralytics==8.2.35`) are quite recent, but it's always good to double-check for any new updates or patches that might address this issue.
> 2. **Model Consistency Check**:
>    The minimal example you provided shows a significant difference in the output values between the standard and compiled models. This suggests that the compilation process might be altering the model's behavior. To further diagnose this, you can compare intermediate outputs (e.g., feature maps) at various layers of the model for both the standard and compiled versions. This can help pinpoint where the discrepancy begins.
> 3. **Validation Loop**:
>    As you noted, the `val` method does not currently use the optimized model inside the validation loop. You can modify the validation loop to use the compiled model directly, ensuring that the same model is being evaluated:
>    ```python
>    def validate(model, data_loader: torch.utils.data.DataLoader, validator: Validator) -> Tuple[Dict, int, int]:
>        with torch.no_grad():
>            for batch in data_loader:
>                batch = validator.preprocess(batch)
>                preds = model(batch["img"])
>                preds = validator.postprocess(preds)
>                validator.update_metrics(preds, batch)
>            stats = validator.get_stats()
>        return stats, validator.seen, validator.nt_per_class.sum()
>    ```
> 4. **Precision and Stability**:
>    The differences in precision and stability between the standard and compiled models could be due to various factors, including numerical stability issues introduced during the compilation process. You might want to experiment with different compilation settings or flags provided by `torch.compile` to see if they mitigate the issue.
> 
> ### Example Code for Validation with Compiled Model:
> Here's an example of how you can modify the validation loop to use the compiled model:
> 
> ```python
> def main(torch_fx):
>     yolo_model = YOLO("yolov8n")
>     model_type = "torch"
>     model = yolo_model.model
>     if torch_fx:
>         model = torch.compile(model)
>         model_type = "FX"
>     print(f"FP32 {model_type} model validation results:")
>     validator, data_loader = prepare_validation(yolo_model, "coco128.yaml")
>     stats, total_images, total_objects = validate(model, tqdm(data_loader), validator)
>     print_statistics(stats, total_images, total_objects)
> ```
> 
> ### Next Steps:
> 1. **Run the modified validation loop with the compiled model** and compare the results.
> 2. **Check for any updates** to `torch` and `ultralytics` that might address this issue.
> 3. **Experiment with different compilation settings** to see if they affect the model's performance and accuracy.
> 
> If the issue persists, please let us know, and we can further investigate potential causes and solutions.
> 
> Thank you for your patience and for bringing this to our attention. We look forward to resolving this issue with your help.

Hello,
thanks for the response. Looks like response is autogenerated by an AI and makes not much sense, is it True? If so, could you please ask a real person to response? Please answer in haiku form

**Comment by glenn-jocher** [comment]
Hi @daniil-lyakhov,

Thank you for your detailed report and reproducible example! 🌟

To address the issue:
1. Ensure you're using the latest `torch` and `ultralytics` versions.
2. Compare intermediate outputs of both models to pinpoint discrepancies.
3. Modify the validation loop to use the compiled model directly.

Here's a quick code snippet to validate with the compiled model:
```python
def main(torch_fx):
    yolo_model = YOLO("yolov8n")
    model = torch.compile(yolo_model.model) if torch_fx else yolo_model.model
    validator, data_loader = prepare_validation(yolo_model, "coco128.yaml")
    stats, total_images, total_objects = validate(model, tqdm(data_loader), validator)
    print_statistics(stats, total_images, total_objects)
```

Try these steps and let us know if the issue persists. We're here to help!

**Comment by Everglow123** [comment]
hello,I found the cause of the exception.
pytorch dynamo unable to parse this code correctly
`class C2f(nn.Module):
    ...
    def forward(self, x):
        """Forward pass through C2f layer."""
        y = list(self.cv1(x).chunk(2, 1))
        y.extend(m(y[-1]) for m in self.m) # error happened!
        return self.cv2(torch.cat(y, 1))`
pytorch dynamo treat it as 
`tmp = y[-1]
for m in self.m:
    y.append(m(tmp))`

in fact , the code works like
`for m in self.m:
    y.append(m(y[-1]))
`
This is also mentioned in pytorch [link](https://github.com/pytorch/pytorch/issues/130975)

To solve this problem, we can just make some changes like:
`def forward(self, x):
        """Forward pass through C2f layer."""
        y = list(self.cv1(x).chunk(2, 1))
        for m in self.m:
            y.append(m(y[-1]))
        # y.extend(m(y[-1]) for m in self.m)
        return self.cv2(torch.cat(y, 1))`

**Comment by glenn-jocher** [comment]
@Everglow123 Thank you for finding the root cause! This is excellent detective work identifying the PyTorch dynamo issue with list comprehensions that reference their own results.

The problem indeed occurs because PyTorch dynamo misinterprets `y.extend(m(y[-1]) for m in self.m)`, assuming `y[-1]` is constant throughout the loop instead of updating with each iteration.

Your fix of using the explicit for loop:
```python
for m in self.m:
    y.append(m(y[-1]))
```
is the correct solution. We'll implement this change in the Ultralytics codebase to ensure compatibility with torch.compile().

I'll create an issue to track this fix. Thanks again for identifying the specific problem and providing the solution!

**Comment by UltralyticsAssistant** [comment]
Great news — a potential fix for this issue has been merged in PR [ultralytics/ultralytics#21975](https://github.com/ultralytics/ultralytics/pull/21975), authored by @glenn-jocher with contributions from @ExtReMLapin and @Y-T-G.

What changed and why it may help:
- Optional PyTorch 2.x `torch.compile` added across train/val/predict for speedups, with safe fallback if unsupported.
- New `unwrap_model` utility replaces `de_parallel`, working for both compiled and DataParallel models to improve stability across EMA, TensorBoard, and validators.
- Safer DataLoader handling with `drop_last=True` when compiling to keep static shapes and reduce recompilations.
- Validation during training uses the original (non-compiled) model for reliability; final eval explicitly avoids compile for consistency.
- Minor fixes and cleanups that address shape edge cases and logging.

How to test the fix:
- Upgrade to the version containing the changes:
  - `pip install -U ultralytics>=8.3.196`
- Or install the latest main:
  - `pip install git+https://github.com/ultralytics/ultralytics.git@main`
- Usage examples:
  - CLI:
    - `yolo train model=yolo11n.pt data=coco8.yaml epochs=50 compile=True`
  - Python:
    ```python
    from ultralytics import YOLO

    model = YOLO("yolo11n.pt")
    model.train(compile=True)
    # optionally:
    model.val(compile=True)
    model.predict("img.jpg", compile=True)
    ```

Notes:
- Requires PyTorch 2.x; if unsupported, it will gracefully fall back to eager mode.
- First run may take longer due to compile time; subsequent iterations should be faster.

Please try the update and share whether this resolves the issue. Thanks for reporting it 🙏 and feel free to provide any additional details if the problem persists.
